src/Microsoft.AspNet.OData.Shared/Query/SkipTokenQueryOption.cs

```diff
@@ -0,0 +1,183 @@
+        /// <summary>
+        /// Gets or sets the list of OrderByNodes for the original query
+        /// </summary>
+        public IList<OrderByNode> OrderByNodes { get; private set; }
+
+        /// <summary>
+        /// Apply the $skiptoken query to the given IQueryable.
+        /// </summary>
+        /// <param name="query">The original <see cref="IQueryable"/>.</param>
+        /// <param name="querySettings">The query settings to use while applying this query option.</param>
+        /// <param name="orderByNodes">Information about the orderby query option.</param>
+        /// <returns>The new <see cref="IQueryable"/> after the skiptoken query has been applied to.</returns>
+        public IQueryable<T> ApplyTo<T>(IQueryable<T> query, ODataQuerySettings querySettings, IList<OrderByNode> orderByNodes)
+        {
+            QuerySettings = querySettings;
+            OrderByNodes = orderByNodes;
+            return SkipTokenHandler.ApplyTo<T>(query, this) as IOrderedQueryable<T>;
+        }
+
+        /// <summary>
+        /// Apply the $skiptoken query to the given IQueryable.
+        /// </summary>
+        /// <param name="query">The original <see cref="IQueryable"/>.</param>
+        /// <param name="querySettings">The query settings to use while applying this query option.</param>
+        /// <param name="orderByNodes">Information about the orderby query option.</param>
+        /// <returns>The new <see cref="IQueryable"/> after the skiptoken query has been applied to.</returns>
+        public IQueryable ApplyTo(IQueryable query, ODataQuerySettings querySettings, IList<OrderByNode> orderByNodes)
+        {
+            QuerySettings = querySettings;
+            OrderByNodes = orderByNodes;
+            return SkipTokenHandler.ApplyTo(query, this);
+        }
+
+        /// <summary>
+        /// Validate the skiptoken query based on the given <paramref name="validationSettings"/>. It throws an ODataException if validation failed.
+        /// </summary>
+        /// <param name="validationSettings">The <see cref="ODataValidationSettings"/> instance which contains all the validation settings.</param>
+        public void Validate(ODataValidationSettings validationSettings)
+        {
+            if (validationSettings == null)
+            {
+                throw Error.ArgumentNull("validationSettings");
+            }
+
+            if (Validator != null)
+            {
+                Validator.Validate(this, validationSettings);
+            }
+        }
+
+        /// <summary>
+        /// Returns the handler for skip token
+        /// </summary>
+        /// <param name="context">The <see cref="ODataQueryContext"/> instance of the query.</param>
+        /// <returns></returns>
+        public static SkipTokenHandler GetSkipTokenHandler(ODataQueryContext context)
+        {
+            if (context == null || context.RequestContainer == null)
+            {
+                return new DefaultSkipTokenHandler();
```

Comment:
>new DefaultSkipTokenHandler() [](start = 23, length = 29)

Do we always create a Default...Handler() ?

Can we create a default and use it for each Request?

Reply:
We only create one if the context or the request container is not accessible. If they exist, we share the handler as the scope is singleton in DI. 

---
In reply to: [281400826](https://github.com/OData/WebApi/pull/1724#discussion_r281400826) [](ancestors = 281400826)